### PR TITLE
[lld] Support merging RISC-V Atomics ABI attributes

### PR DESCRIPTION
**Pull request by ilovepi**
This patch adds support for merging the atomic_abi attribute, specifid in
https://github.com/riscv-non-isa/riscv-elf-psabi-doc/blob/master/riscv-elf.adoc#tag_riscv_atomic_abi-14-uleb128version,
to LLD.

The atomics_abi tag merging is conducted as follows:

UNKNOWN is safe to merge with all other values.
A6C is compatible with A6S, and results in the A6C ABI.
A6C is incompatible with A7, and results in an error.
A6S and A7 are compatible, and merging results in the A7 ABI.
Note: the A7 is not yet supported in either LLVM or in any current hardware,
and is therefore omitted from attribute generation in RISCVTargetStreamer.

LLD support was split from https://github.com/llvm/llvm-project/pull/90266

### CODE REVIEW

File: lld/ELF/Arch/RISCV.cpp [hunk 1]
```diff
@@ -1084,10 +1084,76 @@ static void mergeArch(RISCVISAUtils::OrderedExtensionMap &mergedExts,
...
+
+  auto reportAbiError = [&]() {
+    errorOrWarn("atomic abi mismatch for " + oldSection->name + "\n>>> " +
+                toString(oldSection) +
+                ": atomic_abi=" + Twine(static_cast<unsigned>(oldTag)) +
+                "\n>>> " + toString(newSection) +
+                ": atomic_abi=" + Twine(static_cast<unsigned>(newTag)));
+  };
+
+  switch (static_cast<RISCVAtomicAbiTag>(oldTag)) {
```
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1105]: unneeded cast?

File: lld/ELF/Arch/RISCV.cpp [hunk 2]
```diff
@@ -1084,10 +1084,76 @@ static void mergeArch(RISCVISAUtils::OrderedExtensionMap &mergedExts,
...
+      return;
+    case RISCVAtomicAbiTag::A6C:
+      reportAbiError();
+      return;
+    case RISCVAttrs::RISCVAtomicAbiTag::UNKNOWN:
+    case RISCVAttrs::RISCVAtomicAbiTag::A7:
+      return;
+    };
+  };
+  llvm_unreachable("unknown AtomicABI");
```
Comment (MaskRay) [inline lld/ELF/Arch/RISCV.cpp:1148]: If a potentially-corrupted object file may contain an unknown tag, it's inappropriate to use unreachable.

File: lld/ELF/Arch/RISCV.cpp (hunk 1 above)
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1105]: yes. Looks like I missed one. Thanks.

File: lld/ELF/Arch/RISCV.cpp (hunk 2 above)
Comment (ilovepi) [reply lld/ELF/Arch/RISCV.cpp:1148]: Ah, fair point. I'll update w/ an error instead.

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
> [RISCV][lld]  ...

I usually omit `[RISCV]` when the title already contains `RISC-V` or `RISCV`...

**Pull request merged**